src/Umbraco.Infrastructure/Security/MemberRoleStore.cs

```diff
@@ -0,0 +1,342 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Services;
+
+namespace Umbraco.Cms.Core.Security
+{
+    /// <summary>
+    /// A custom user store that uses Umbraco member data
+    /// </summary>
+    public class MemberRoleStore<TRole> : IRoleStore<TRole> where TRole : IdentityRole
+    {
+        private readonly IMemberGroupService _memberGroupService;
+        private bool _disposed;
+
+        //TODO: Move into custom error describer.
+        //TODO: How revealing can the error messages be?
+        private readonly IdentityError _intParseError = new IdentityError { Code = "IdentityIdParseError", Description = "Cannot parse ID to int" };
+        private readonly IdentityError _memberGroupNotFoundError = new IdentityError { Code = "IdentityMemberGroupNotFound", Description = "Member group not found" };
+        private const string genericIdentityErrorCode = "IdentityErrorUserStore";
+
+        public MemberRoleStore(IMemberGroupService memberGroupService, IdentityErrorDescriber errorDescriber)
+        {
+            _memberGroupService = memberGroupService ?? throw new ArgumentNullException(nameof(memberGroupService));
+            ErrorDescriber = errorDescriber ?? throw new ArgumentNullException(nameof(errorDescriber));
+        }
+
+        /// <summary>
+        /// Gets or sets the <see cref="IdentityErrorDescriber"/> for any error that occurred with the current operation.
+        /// </summary>
+        public IdentityErrorDescriber ErrorDescriber { get; set; }
+
+        /// <inheritdoc />
+        public Task<IdentityResult> CreateAsync(TRole role, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                ThrowIfDisposed();
+
+                if (role == null)
+                {
+                    throw new ArgumentNullException(nameof(role));
+                }
+
+                var memberGroup = new MemberGroup
+                {
+                    Name = role.Name
+                };
+
+                _memberGroupService.Save(memberGroup);
+
+                role.Id = memberGroup.Id.ToString();
+
+                return Task.FromResult(IdentityResult.Success);
+            }
+            catch (Exception ex)
+            {
+                return Task.FromResult(IdentityResult.Failed(new IdentityError { Code = genericIdentityErrorCode, Description = ex.Message }));
+            }
+        }
+
+
+        /// <inheritdoc />
+        public Task<IdentityResult> UpdateAsync(TRole role, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                if (role == null)
+                {
+                    throw new ArgumentNullException(nameof(role));
+                }
+
+                ThrowIfDisposed();
```

Comment:
**nitpick:** the other methods check for disposed before null role

Reply:
Good spot, amended